src/k8s/pkg/k8sd/controllers/node_configuration.go

```diff
@@ -14,24 +14,48 @@ import (
 )
 
 type NodeConfigurationController struct {
-	snap            snap.Snap
-	createK8sClient func(ctx context.Context) *k8s.Client
+	snap         snap.Snap
+	waitReady    func()
+	newK8sClient func() (*k8s.Client, error)
 }
 
-func NewNodeConfigurationController(snap snap.Snap, createK8sClient func(ctx context.Context) *k8s.Client) *NodeConfigurationController {
+func NewNodeConfigurationController(snap snap.Snap, waitReady func(), newK8sClient func() (*k8s.Client, error)) *NodeConfigurationController {
 	return &NodeConfigurationController{
-		snap:            snap,
-		createK8sClient: createK8sClient,
+		snap:         snap,
+		waitReady:    waitReady,
+		newK8sClient: newK8sClient,
+	}
+}
+
+func (c *NodeConfigurationController) retryNewK8sClient(ctx context.Context) (*k8s.Client, error) {
+	for {
+		client, err := c.newK8sClient()
+		if err == nil {
+			return client, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(3 * time.Second):
+		}
 	}
 }
 
 func (c *NodeConfigurationController) Run(ctx context.Context) {
-	client := c.createK8sClient(ctx)
+	// wait for microcluster node to be ready
+	c.waitReady()
+
+	client, err := c.retryNewK8sClient(ctx)
```

Comment:
why does the k8s client might fail if the cluster is ready?

Reply:
database is open, but could be that we hit a race condition where kubeconfig files are not yet in place